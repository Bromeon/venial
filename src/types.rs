#![allow(missing_docs)]

use proc_macro2::{Ident, Literal, TokenStream, TokenTree};
use quote::{ToTokens, TokenStreamExt as _};

// TODO - handle unions

/// The declaration of a Rust type.
///
/// **Example input:**
///
/// ```no_run
/// struct MyUnitStruct;
/// struct MyTupleStruct(i32, f32);
/// struct MyRegularStruct {
///     // ...
/// }
/// enum MyEnum {
///     // ...
/// }
/// ```
#[non_exhaustive]
#[derive(Clone, Debug)]
pub enum Declaration {
    Struct(Struct),
    Enum(Enum),
    Union(Union),
    Function(Function),
}

// TODO - fn Declaration::name()
// TODO - fn Declaration::as_struct()
// TODO - fn Declaration::as_enum()

/// Declaration of a struct.
///
/// **Example input:**
///
/// ```no_run
/// struct MyUnitStruct;
/// struct MyTupleStruct(i32, f32);
/// struct MyRegularStruct {
///     // ...
/// }
/// ```
#[derive(Clone, Debug)]
pub struct Struct {
    pub attributes: Vec<Attribute>,
    pub vis_marker: Option<VisMarker>,
    pub name: Ident,
    pub generic_params: Option<GenericParams>,
    pub where_clauses: Option<WhereClauses>,
    pub fields: StructFields,
}

// TODO - fn Struct::field_names()
// TODO - fn Struct::field_types()

/// Fields of a [`Struct`] or an [`EnumVariant`].
#[derive(Clone, Debug)]
pub enum StructFields {
    Unit,
    Tuple(Vec<TupleField>),
    Named(Vec<NamedField>),
}

/// Declaration of an enum.
///
/// **Example input:**
///
/// ```no_run
/// enum MyEnum {
///     // ...
/// }
/// ```
#[derive(Clone, Debug)]
pub struct Enum {
    pub attributes: Vec<Attribute>,
    pub vis_marker: Option<VisMarker>,
    pub name: Ident,
    pub generic_params: Option<GenericParams>,
    pub where_clauses: Option<WhereClauses>,
    pub variants: Vec<EnumVariant>,
}

// TODO - fn Enum::is_c_enum()

/// The individual variant of an [`Enum`].
///
/// The variant can either be a c-like variant, hold one or multiple types,
/// or hold named fields depending on the value of `contents`.
#[derive(Clone, Debug)]
pub struct EnumVariant {
    pub attributes: Vec<Attribute>,
    pub vis_marker: Option<VisMarker>,
    pub name: Ident,
    pub contents: StructFields,

    /// The value of the variant, normally for c-like enums.
    pub discriminant: Option<EnumDiscriminant>,
}

// TODO - fn EnumVariant::is_empty_variant()
// TODO - fn EnumVariant::get_single_type()

/// Declaration of an union.
///
/// **Example input:**
///
/// ```no_run
/// union MyUnion {
///     // ...
/// }
/// ```
#[derive(Clone, Debug)]
pub struct Union {
    pub attributes: Vec<Attribute>,
    pub vis_marker: Option<VisMarker>,
    pub name: Ident,
    pub generic_params: Option<GenericParams>,
    pub where_clauses: Option<WhereClauses>,
    pub fields: Vec<NamedField>,
}

/// Declaration of a function.
///
/// **Example input:**
///
/// ```no_run
/// const fn hello(a: i32, b: f32) -> f32 { return 0.0; }
/// unsafe fn eval(c: String, b: i32) { return; }
/// fn do_thing<T: Clone>(t: T) where T: Default;
/// ```
#[derive(Clone, Debug)]
pub struct Function {
    pub attributes: Vec<Attribute>,
    pub vis_marker: Option<VisMarker>,
    pub qualifiers: FunctionQualifiers,
    pub name: Ident,
    pub generic_params: Option<GenericParams>,
    pub params: Vec<FunctionParameter>,
    pub where_clauses: Option<WhereClauses>,
    pub return_ty: Option<TyExpr>,
    pub body: Option<TokenTree>,
}

/// Keywords giving special information on a function.
///
/// Possible qualifiers are `default`, `const`, `async`, `unsafe` and `extern`,
/// always in that order.
#[derive(Clone, Debug, Default)]
pub struct FunctionQualifiers {
    pub is_default: bool,
    pub is_const: bool,
    pub is_async: bool,
    pub is_unsafe: bool,
    pub is_extern: bool,
    pub extern_abi: Option<Literal>,
}

/// A parameter of a [`Function`]
///
/// In the following code, the function parameters captured are `a: i32` and `b: f32`
///
/// ```no_run
/// pub fn hello_world(a: i32, b: f32) {}
/// ```
#[derive(Clone, Debug)]
pub struct FunctionParameter {
    pub attributes: Vec<Attribute>,
    pub name: Ident,
    pub ty: TyExpr,
}

/// A field of a tuple [`Struct`] or tuple-like [`EnumVariant`].
///
/// For instance, in `struct MyTuple(A, B, C);` A, B, C are each a tuple field.
#[derive(Clone, Debug)]
pub struct TupleField {
    pub attributes: Vec<Attribute>,
    pub vis_marker: Option<VisMarker>,
    pub ty: TyExpr,
}

/// A field of a [`Struct`] or struct-like [`EnumVariant`].
///
/// For instance, in the following code, `foo` and `bar` are each a struct field:
///
/// ```no_run
/// struct MyStruct {
///     foo: i32,
///     bar: f32,
/// }
/// ```
#[derive(Clone, Debug)]
pub struct NamedField {
    pub attributes: Vec<Attribute>,
    pub vis_marker: Option<VisMarker>,
    pub name: Ident,
    pub ty: TyExpr,
}

// --- Token groups ---

// TODO - parse better
// see https://doc.rust-lang.org/reference/attributes.html

/// An outer attribute.
///
/// **Example input:**
///
/// ```no_run
/// # #[cfg(FALSE)]
/// #[hello(world)]
/// # struct Foo;
/// ```
#[derive(Clone)]
pub struct Attribute {
    pub _hashbang: TokenTree,
    pub child_tokens: Vec<TokenTree>,
}

/// Visibility marker, eg `pub`, `pub(crate)`, `pub(super)`, etc.
#[derive(Clone)]
pub struct VisMarker {
    /// `pub` token.
    pub _token1: TokenTree,
    /// `(...)` token, if any.
    pub _token2: Option<TokenTree>,
}

/// The generic parameters declared right after your type's name.
///
/// **Example input:**
///
/// ```no_run
/// struct MyUnitStruct<A, B, C>(A, B, C);
/// ```
#[derive(Clone)]
pub struct GenericParams {
    pub tokens: Vec<TokenTree>,
}

/// All the stuff that comes after the `where` keyword.
#[derive(Clone)]
pub struct WhereClauses {
    pub tokens: Vec<TokenTree>,
}

/// Type expression in a [`TupleField`] or [`NamedField`].
///
/// **Example input:**
///
/// ```no_run
/// # struct MyUnitStruct<'a> {
///     foo: i32,
///     bar: &'a mut Vec<(u64, Option<bool>, f64)>,
/// };
/// ```
#[derive(Clone)]
pub struct TyExpr {
    pub tokens: Vec<TokenTree>,
}

/// The value of an [`EnumVariant`], normally for c-like enums.
///
/// **Example input:**
///
/// ```no_run
/// enum MyEnum {
///     A = 42,
///     # #[cfg(FALSE)]
///     B = (some + arbitrary.expression()),
/// }
/// ```
#[derive(Clone)]
pub struct EnumDiscriminant {
    pub tokens: Vec<TokenTree>,
}

// ---

impl std::fmt::Debug for Attribute {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("#")?;
        let mut list = f.debug_list();
        for token in &self.child_tokens {
            match token {
                TokenTree::Group(_group) => list.entry(token),
                TokenTree::Ident(_ident) => list.entry(&token.to_string()),
                TokenTree::Punct(_punct) => list.entry(&token.to_string()),
                TokenTree::Literal(_literal) => list.entry(&token.to_string()),
            };
        }
        list.finish()
    }
}

impl std::fmt::Debug for VisMarker {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self._token2 {
            None => f.write_str(&self._token1.to_string()),
            Some(TokenTree::Group(group)) => {
                let mut list = f.debug_tuple(&self._token1.to_string());
                for token in group.stream() {
                    match &token {
                        TokenTree::Group(_group) => list.field(&token),
                        TokenTree::Ident(_ident) => list.field(&token.to_string()),
                        TokenTree::Punct(_punct) => list.field(&token.to_string()),
                        TokenTree::Literal(_literal) => list.field(&token.to_string()),
                    };
                }
                list.finish()
            }
            _ => unreachable!(),
        }
    }
}

impl std::fmt::Debug for GenericParams {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut list = f.debug_list();
        for token in &self.tokens {
            match token {
                TokenTree::Group(_group) => list.entry(token),
                TokenTree::Ident(_ident) => list.entry(&token.to_string()),
                TokenTree::Punct(_punct) => list.entry(&token.to_string()),
                TokenTree::Literal(_literal) => list.entry(&token.to_string()),
            };
        }
        list.finish()
    }
}

impl std::fmt::Debug for WhereClauses {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut list = f.debug_list();
        for token in &self.tokens {
            match token {
                TokenTree::Group(_group) => list.entry(token),
                TokenTree::Ident(_ident) => list.entry(&token.to_string()),
                TokenTree::Punct(_punct) => list.entry(&token.to_string()),
                TokenTree::Literal(_literal) => list.entry(&token.to_string()),
            };
        }
        list.finish()
    }
}

impl std::fmt::Debug for TyExpr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut list = f.debug_list();
        for token in &self.tokens {
            match token {
                TokenTree::Group(_group) => list.entry(token),
                TokenTree::Ident(_ident) => list.entry(&token.to_string()),
                TokenTree::Punct(_punct) => list.entry(&token.to_string()),
                TokenTree::Literal(_literal) => list.entry(&token.to_string()),
            };
        }
        list.finish()
    }
}

impl std::fmt::Debug for EnumDiscriminant {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut list = f.debug_list();
        for token in &self.tokens {
            match token {
                TokenTree::Group(_group) => list.entry(token),
                TokenTree::Ident(_ident) => list.entry(&token.to_string()),
                TokenTree::Punct(_punct) => list.entry(&token.to_string()),
                TokenTree::Literal(_literal) => list.entry(&token.to_string()),
            };
        }
        list.finish()
    }
}

impl ToTokens for Attribute {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        tokens.append(self._hashbang.clone());
        for token in &self.child_tokens {
            tokens.append(token.clone());
        }
    }
}

impl ToTokens for VisMarker {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        self._token1.to_tokens(tokens);
        self._token2.to_tokens(tokens);
    }
}

impl ToTokens for GenericParams {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        for token in &self.tokens {
            tokens.append(token.clone());
        }
    }
}

impl ToTokens for WhereClauses {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        for token in &self.tokens {
            tokens.append(token.clone());
        }
    }
}

impl ToTokens for TyExpr {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        for token in &self.tokens {
            tokens.append(token.clone());
        }
    }
}

impl ToTokens for EnumDiscriminant {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        for token in &self.tokens {
            tokens.append(token.clone());
        }
    }
}
