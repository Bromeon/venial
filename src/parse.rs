use crate::types::*;
use proc_macro2::{token_stream, Delimiter, Ident, TokenStream, TokenTree};

fn parse_ident(token: TokenTree) -> Option<Ident> {
    match token {
        TokenTree::Group(_group) => None,
        TokenTree::Ident(ident) => Some(ident),
        TokenTree::Punct(_punct) => None,
        TokenTree::Literal(_literal) => None,
    }
}

fn consume_until_period(
    first_token: Option<TokenTree>,
    tokens: &mut token_stream::IntoIter,
) -> Vec<TokenTree> {
    let mut tokens_before_period = Vec::new();

    if let Some(token) = first_token {
        tokens_before_period.push(token)
    }

    for token in tokens.into_iter() {
        match token {
            TokenTree::Punct(punct) if punct.as_char() == ',' => break,
            _ => {
                tokens_before_period.push(token);
            }
        }
    }

    tokens_before_period
}

pub fn parse_tuple_fields(tokens: TokenStream) -> Vec<TupleField> {
    // TODO - attributes
    // TODO - skip generic params
    let mut fields = Vec::new();

    let mut tokens = tokens.into_iter();
    loop {
        let next_token = if let Some(next_token) = tokens.next() {
            next_token
        } else {
            break;
        };

        fields.push(TupleField {
            ty: TyExpr {
                tokens: consume_until_period(Some(next_token), &mut tokens),
            },
        });
    }

    fields
}

pub fn parse_named_fields(tokens: TokenStream) -> Vec<NamedField> {
    // TODO - attributes
    // TODO - skip generic params
    let mut fields = Vec::new();

    let mut tokens = tokens.into_iter();
    loop {
        let next_token = if let Some(next_token) = tokens.next() {
            next_token
        } else {
            break;
        };

        let ident = parse_ident(next_token).unwrap();

        match tokens.next() {
            Some(TokenTree::Punct(punct)) if punct.as_char() == ':' => (),
            _ => panic!("cannot parse type"),
        };

        let next_token = tokens.next().unwrap();
        fields.push(NamedField {
            name: ident.to_string(),
            ty: TyExpr {
                tokens: consume_until_period(Some(next_token), &mut tokens),
            },
        });
    }

    fields
}

pub fn parse_type(tokens: TokenStream) -> TypeDeclaration {
    let mut tokens = tokens.into_iter();

    let next_token = tokens.next().unwrap();

    if let Some(ident) = parse_ident(next_token) {
        if ident == "struct" {
            let next_token = tokens.next().unwrap();
            let struct_name = parse_ident(next_token).unwrap();

            let next_token = tokens.next().unwrap();
            let struct_fields = match next_token {
                TokenTree::Punct(punct) if punct.as_char() == ';' => StructFields::Unit,
                TokenTree::Group(group) if group.delimiter() == Delimiter::Parenthesis => {
                    StructFields::Tuple(parse_tuple_fields(group.stream()))
                }
                TokenTree::Group(group) if group.delimiter() == Delimiter::Brace => {
                    StructFields::Named(parse_named_fields(group.stream()))
                }
                _ => panic!("cannot parse type"),
            };

            return TypeDeclaration::Struct(Struct {
                name: struct_name.to_string(),
                contents: struct_fields,
            });
        } else if ident == "enum" {
            todo!()
        } else {
            panic!("cannot parse type")
        }
    } else {
        panic!("cannot parse type")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use quote::quote;

    #[test]
    fn parse_unit_struct() {
        let struct_type = parse_type(quote!(
            struct Hello;
        ));

        assert_debug_snapshot!(struct_type);
    }

    #[test]
    fn parse_tuple_struct() {
        let struct_type = parse_type(quote!(
            struct Hello(A, B);
        ));

        assert_debug_snapshot!(struct_type);
    }

    #[test]
    fn parse_normal_struct() {
        let struct_type = parse_type(quote!(
            struct Hello {
                a: A,
                b: B,
            }
        ));

        assert_debug_snapshot!(struct_type);
    }
}
