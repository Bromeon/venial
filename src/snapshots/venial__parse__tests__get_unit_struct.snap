---
source: src/parse.rs
assertion_line: 138
expression: struct_type
---
Struct(
    Struct {
        name: "Hello",
        contents: Unit,
    },
)
